dbms/src/Server/ServerInfo.cpp

```diff
@@ -0,0 +1,184 @@
+void ServerInfo::parseDiskInfo(const diagnosticspb::ServerInfoItem & disk_info_item)
+{
+    Disk disk;
+    disk.name = disk_info_item.name();
+    for (const auto & pair : disk_info_item.pairs())
+    {
+        if (pair.key() == "type")
+        {
+            std::unordered_map<String, Disk::DiskType> disk_type_map = {
+                {"hdd", Disk::DiskType::HDD},
+                {"ssd", Disk::DiskType::SSD},
+                {"unknown", Disk::DiskType::UNKNOWN}};
+            disk.disk_type = disk_type_map[pair.value()];
+        }
+        else if (pair.key() == "total")
+        {
+            disk.total_space = static_cast<UInt64>(std::stoull(pair.value()));
+        }
+        else if (pair.key() == "free")
+        {
+            disk.free_space = static_cast<UInt64>(std::stoull(pair.value()));
+        }
+        else if (pair.key() == "path")
+        {
+            disk.mount_point = pair.value();
+        }
+        else if (pair.key() == "fstype")
+        {
+            disk.fs_type = pair.value();
+        }
+    }
+    disk_infos.push_back(disk);
+}
+
+void ServerInfo::parseMemoryInfo(const diagnosticspb::ServerInfoItem & memory_info_item)
+{
+    for (const auto & pair : memory_info_item.pairs())
+    {
+        if (pair.key() == "capacity")
+        {
+            memory_info.capacity = std::stoull(pair.value());
+        }
+    }
+}
+
+void ServerInfo::parseSysInfo(diagnosticspb::ServerInfoResponse & sys_info_response)
+{
+    for (const auto & item : sys_info_response.items())
+    {
+        if (item.tp() == "cpu")
+        {
+            parseCPUInfo(item);
+        }
+        else if (item.tp() == "disk")
+        {
+            parseDiskInfo(item);
+        }
+        else if (item.tp() == "memory")
+        {
+            parseMemoryInfo(item);
+        }
+    }
+}
+
+String ServerInfo::debugString() const
+{
+    FmtBuffer fmt_buf;
+    fmt_buf.fmtAppend("CPU: \n"
+                      "     logical_cores: {}\n"
+                      "     physical_cores: {}\n"
+                      "     frequency: {}\n"
+                      "     l1_cache_size: {}\n"
+                      "     l1_cache_line_size: {}\n"
+                      "     l2_cache_size: {}\n"
+                      "     l2_cache_line_size: {}\n"
+                      "     l3_cache_size: {}\n"
+                      "     l3_cache_line_size: {}\n"
+                      "     arch: {}\n",
+                      cpu_info.logical_cores,
+                      cpu_info.physical_cores,
+                      cpu_info.frequency,
+                      cpu_info.l1_cache_size,
+                      cpu_info.l1_cache_line_size,
+                      cpu_info.l2_cache_size,
+                      cpu_info.l2_cache_line_size,
+                      cpu_info.l3_cache_size,
+                      cpu_info.l3_cache_line_size,
+                      cpu_info.arch);
+
+    for (const auto & disk_info : disk_infos)
+    {
+        fmt_buf.fmtAppend("Disk: \n"
```

Comment:
The `joinStr` grammar is too complex like below and not efficient here, so I think it may be better to stand.
```cpp
fmt_buf.joinStr(
        disk_infos.cbegin(),
        disk_infos.cend(),
        [](const Disk & disk, FmtBuffer & fb) {
            fb.fmtAppend("Disk: \n"
                         "     name: {}\n"
                         "     type: {}\n"
                         "     total_space: {}\n"
                         "     free_space: {}\n"
                         "     mount_point: {}\n"
                         "     fs_type: {}\n",
                         disk.name,
                         disk.disk_type,
                         disk.total_space,
                         disk.free_space,
                         disk.mount_point,
                         disk.fs_type);
        },
        "");
```